feat: Meeting Schedule Quickstart

zepfred opened the pull request:
## Description of the change

This PR adds a new quickstart for the Meeting Scheduling problem.

[Screencast from 2024-04-08 13-07-14.webm](https://github.com/TimefoldAI/timefold-quickstarts/assets/11961659/c84e91cc-5deb-4c3e-b1b1-c6bdb0705035)

## Checklist

### Development

- [x] The changes have been covered with tests, if necessary.
- [x] You have a green build, with the exception of the flaky tests.
- [x] UI and JS files are fully tested, the user interface works for all modules affected by your changes (e.g., solve and analyze buttons).
- [x] The network calls work for all modules affected by your changes (e.g., solving a problem).
- [x] The console messages are validated for all modules affected by your changes.

### Code Review

- [x] This pull request includes an explanatory title and description.
- [x] The GitHub issue is linked.
- [x] At least one other engineer has approved the changes.
- [ ] After PR is merged, inform the reporter.

triceo commented on use-cases/meeting-scheduling/pom.xml:
```diff
@@ -0,0 +1,229 @@
+        <maven.compiler.release>17</maven.compiler.release>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+
+        <version.io.quarkus>3.9.1</version.io.quarkus>
```
We're back to 3.8.3 for now.

triceo commented on use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/rest/DemoDataGenerator.java:
```diff
@@ -0,0 +1,255 @@
+
+import jakarta.enterprise.context.ApplicationScoped;
+
+import ai.timefold.solver.core.impl.util.MutableInt;
+import ai.timefold.solver.core.impl.util.Pair;
```
Let's not use internal classes from the solver.
If you need a `Pair`, write the record yourself - it's trivial.
The `MutableInt` is unnecessary.

triceo commented on use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/solver/MeetingSchedulingConstraintProvider.java:
```diff
@@ -0,0 +1,224 @@
+                                (attendee1, attendee2, assignment) -> assignment.getStartingTimeGrain().getGrainIndex() +
+                                        assignment.getMeeting().getDurationInGrains(),
+                                assignment -> assignment.getStartingTimeGrain().getGrainIndex(),
+                                assignment -> assignment.getStartingTimeGrain().getGrainIndex() +
```
the grainIndex+durationInGrains is repeated so often that I think it would warrant a getter on assignment.

triceo reviewed (commented):
First quick look.

triceo commented on use-cases/bed-allocation/src/main/java/org/acme/bedallocation/rest/DemoDataGenerator.java:
```diff
@@ -333,4 +331,51 @@ private <T, L> void applyRandomValue(int count, List<T> values, L secondParam, P
+        }
+    }
+
+    private final class MutableReference<Value_> {
+        private Value_ value;
+
+        public MutableReference(Value_ value) {
+            this.value = value;
+        }
+
+        public Value_ getValue() {
+            return this.value;
+        }
+
+        public void setValue(Value_ value) {
+            this.value = value;
+        }
+
+        public boolean equals(Object o) {
+            if (o instanceof MutableReference<?> other) {
+                return this.value.equals(other.value);
+            } else {
+                return false;
+            }
+        }
+
+        public int hashCode() {
+            return Objects.hash(this.value);
+        }
+
+        public String toString() {
+            return this.value.toString();
+        }
+    }
```
Considering that this is a quickstart, let's refactor the code so that mutable references aren't necessary. Every time you have to ask yourself "why is this here?" and the answer is not directly related to the idea you want to show the users, remove the thing.

triceo commented on use-cases/bed-allocation/src/main/java/org/acme/bedallocation/rest/DemoDataGenerator.java:
```diff
@@ -333,4 +331,51 @@ private <T, L> void applyRandomValue(int count, List<T> values, L secondParam, P
+            this.value = value;
+        }
+
+        public Key_ key() {
+            return this.key;
+        }
+
+        public Value_ value() {
+            return this.value;
+        }
```
It's a record; these are not necessary.

triceo commented on use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/Meeting.java:
```diff
@@ -0,0 +1,135 @@
+        return requiredAttendances.size() + preferredAttendances.size();
+    }
+
+    public void addAttendant(String id, Person person, boolean isRequired) {
+        if (isRequired) {
+            if (requiredAttendances.stream().noneMatch(r -> r.getPerson().equals(person))) {
+                requiredAttendances.add(new RequiredAttendance(id, this, person));
+            }
+        } else {
+            if (preferredAttendances.stream().noneMatch(r -> r.getPerson().equals(person))) {
+                preferredAttendances.add(new PreferredAttendance(id, this, person));
+            }
+        }
+    }
```
I see two options here:
- Either don't check if the person is already present.
- Or if you want to check for that, do something if the person is already present. Fail, perhaps.

Also, instead of this if switch, I'd rather turn this into two methods - add required attendant, add preferred attendant. See how the `isRequired` method parameter effectively makes it two different methods? There is no shared code inside of the method.

triceo commented on use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/MeetingAssignment.java:
```diff
@@ -0,0 +1,152 @@
+        // start is inclusive, end is exclusive
+        int start = startingTimeGrain.getGrainIndex();
+        int otherStart = other.startingTimeGrain.getGrainIndex();
+        int otherEnd = otherStart + other.meeting.getDurationInGrains();
```
Didn't we add a method for this?

zepfred replied:
No, we did not.

triceo commented on use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/MeetingAssignment.java:
```diff
@@ -0,0 +1,152 @@
+        if (otherEnd < start) {
+            return 0;
+        }
+        int end = start + meeting.getDurationInGrains();
```
Dtto.

triceo commented on use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/MeetingAssignment.java:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    @JsonIgnore
+    public Integer getLastTimeGrainIndex() {
```
Arguably if the counterpart to this is called `getStarting...`, this one should be `getEnding...`. Or they should be first/last. But it should be mutually consistent.

triceo commented on use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/TimeGrain.java:
```diff
@@ -0,0 +1,134 @@
+@JsonIdentityInfo(scope = TimeGrain.class, generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
+public class TimeGrain implements Comparable<TimeGrain> {
+
+    private static final Comparator<TimeGrain> COMPARATOR = Comparator.comparing(TimeGrain::getDayOfYear)
+            .thenComparingInt(TimeGrain::getStartingMinuteOfDay);
```
Why is `TimeGrain` comparable and can we do anything to stop it?

zepfred replied:
We can remove it by using a `LocalDateTime`.

triceo replied:
Hmm, if you say we need it for joiners in constraints, then I'm fine with `TimeGrain` being comparable. (Better do it directly instead of exposing local date and doing it there.) I was just making sure we're not carrying over some old cruft.

zepfred replied:
Let me undo the change.

triceo commented on use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/TimeGrain.java:
```diff
@@ -0,0 +1,134 @@
+        return LocalTime.of(startingMinuteOfDay / 60, startingMinuteOfDay % 60);
+    }
+
+    @JsonIgnore
+    public String getTimeString() {
+        int hourOfDay = startingMinuteOfDay / 60;
+        int minuteOfHour = startingMinuteOfDay % 60;
+        return (hourOfDay < 10 ? "0" : "") + hourOfDay
+                + ":" + (minuteOfHour < 10 ? "0" : "") + minuteOfHour;
+    }
+
+    @JsonIgnore
+    public String getDateTimeString() {
+        return DAY_FORMATTER.format(getDate()) + " " + getTimeString();
+    }
```
Do we need this anywhere? What for?

zepfred replied:
Good point. It is only used for `TimeGrain::toString()` but can be removed.

triceo commented on use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/rest/DemoDataGenerator.java:
```diff
@@ -0,0 +1,268 @@
+        meetings.stream()
+                .filter(m -> m.getDurationInGrains() == 0)
+                .forEach(m -> m.setDurationInGrains(8));
+        // Attendants
+        MutableInt attendantCount = new MutableInt();
```
If you don't use lambdas so much, you won't have to complicate your code with mutable references.

triceo commented on use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/rest/DemoDataGenerator.java:
```diff
@@ -0,0 +1,268 @@
+            this.value = value;
+        }
+
+        public Key_ key() {
+            return this.key;
+        }
+
+        public Value_ value() {
+            return this.value;
+        }
```
Records don't need these.

triceo commented on use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/solver/MeetingSchedulingConstraintProvider.java:
```diff
@@ -0,0 +1,224 @@
+        return constraintFactory.forEachUniquePair(MeetingAssignment.class,
+                equal(MeetingAssignment::getRoom),
+                overlapping(MeetingAssignment::getGrainIndex,
+                        assignment -> assignment.getGrainIndex() +
+                                assignment.getMeeting().getDurationInGrains()))
```
Don't we have a method for this? If not, let's add it. It's used very often. Please check the rest of the constraint provider too.

zepfred replied:
We have a slightly different method for calculating the index. Now, we have a method for calculating the duration.

triceo reviewed (commented):
Nice!

Besides the comments in-line, I suggest we slightly reduce the number of people in the list. It seems a bit too long once again.

triceo commented on use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/domain/MeetingAssignment.java:
```diff
@@ -99,23 +105,30 @@ public int calculateOverlap(MeetingAssignment other) {
+            return 0;
+        }
+        return startingTimeGrain.getGrainIndex() + meeting.getDurationInGrains();
     }
```
Now this is confusing.

`lastTimeGrainIndex` makes perfect sense - it is the last index. But `endingTimeGrainDuration` is also a grain index; arguably it's not a duration at all, because it starts from zero, not from the starting time grain.

The difference between the two is effectively `-1`. And that, arguably, isn't worth the confusion, or worth us sitting down and figuring out a better name for the method. Therefore my proposal:

- Keep `lastTimeGrainIndex` as that's consistent with `firstTimeGrainIndex`.
- Make `endingTimeGrainDuration` be `lastTimeGrainIndex + 1`.
- Inline all uses of `endingTimeGrainDuration`, removing the method entirely. Can be done with IntelliJ refactoring tools.

triceo reviewed (approved):
One more inline comment. (Sorry if I've changed my mind; I honestly didn't foresee the consequences of my original ask.)

Besides that:

- The red line is still there in the visualization.
- Any chance we could only display "reasonable" business hours in the view? Say 8:00 to 20:00? If not, nevermind.

After that, LGTM.

zepfred commented:
> One more inline comment. (Sorry if I've changed my mind; I honestly didn't foresee the consequences of my original ask.)
> 
> Besides that:
> 
> * The red line is still there in the visualization.

Fixed.

> * Any chance we could only display "reasonable" business hours in the view? Say 8:00 to 20:00? If not, nevermind.

Done.
